geopandas/sindex.py

```diff
@@ -1,25 +1,151 @@
+    if compat.HAS_RTREE:
+        return RTreeIndex
+    warn("Spatial indexes require either `rtree` or `pygeos`.")
+    return None
+
+
+if compat.HAS_RTREE:
+
+    import rtree.index  # noqa
+    from rtree.core import RTreeError  # noqa
+
+    class SpatialIndex(rtree.index.Index):
+        """
+        Original rtree wrapper, kept for backwards compatibility.
+        """
+
+        def __init__(self, *args):
+            super().__init__(self, *args)
+
+        @property
+        def size(self):
+            return len(self.leaves()[0][1])
+
+        @property
+        def is_empty(self):
+            if len(self.leaves()) > 1:
+                return False
+            return self.size < 1
+
+    class RTreeIndex(rtree.index.Index):
+        """
+        A simple wrapper around rtree's RTree Index
+        """
+
+        def __init__(self, geometry):
+            stream = (
+                (i, item.bounds, idx)
+                for i, (idx, item) in enumerate(geometry.iteritems())
+                if pd.notnull(item) and not item.is_empty
+            )
+            try:
+                super().__init__(stream)
+            except RTreeError:
+                # What we really want here is an empty generator error, or
+                # for the bulk loader to log that the generator was empty
+                # and move on.
+                # See https://github.com/Toblerity/rtree/issues/20.
+                super().__init__()
+
+        @property
+        def size(self):
+            return len(self.leaves()[0][1])
+
+        @property
+        def is_empty(self):
+            return self.size == 0
+
+
+if compat.HAS_PYGEOS:
+
+    from pygeos import STRtree, box, points  # noqa
+
+    class PyGEOSSTRTreeIndex(STRtree):
+        """
+        A simple wrapper around pygeos's STRTree
+        """
+
+        with_objects = namedtuple("with_objects", "object id")
+
+        def __init__(self, geometry):
+            # for compatibility with old RTree implementation, store ids/indexes
+            original_indexes = geometry.index
+            non_empty = geometry[~geometry.values.is_empty]
+            self.objects = self.ids = original_indexes[~geometry.values.is_empty]
+            super().__init__(non_empty.values.data)
+
+        def intersection(self, coordinates, objects=False):
+            """Wrapper for pygeos.query that uses the RTree API.
+
+            Parameters
+            ----------
+            coordinates : sequence or array
+                Sequence of the form (min_x, min_y, max_x, max_y)
+                to query a rectangle or (x, y) to query a point.
+            objects : True or False
+                If True, return the label based indexes. If False, integer indexes
+                are returned.
+            """
+            # convert bounds to geometry
+            # the old API uses tuples of bound, but pygeos uses geometries
+            try:
+                iter(coordinates)
+            except TypeError:
```

Comment:
Instead of this try/except, could also do a `if not isinstance(coordinates, collections.abc.Iterable)`

Reply:
The reason I did it like this is to mirror what `rtree` does ([source](https://github.com/Toblerity/rtree/blob/98c2ecffcccd8565b44f631d52969be4b25c0f84/rtree/index.py#L332)), but I do think what you're proposing is better. I think the result would be the same in all cases, but I'm not completely sure.